Updating monitors to colocate to primal grid, and colocate=True by default

momchil-flex opened the pull request:
[backend PR](https://github.com/flexcompute/tidy3d-core/pull/323)

From a user point of view, the most notable changes of this PR are as described in the title:

- FieldTimeMonitor-s and FieldMonitor-s that have `colocate=True` return fields colocated to the grid *boundaries* rather than centers.
- `colocate` = True is now set by default.
- The `colocate` argument was also introduced in the `ModeSolver` and `ModeSolverMonitor`, also set to True by default.

The goal of this is to avoid some common confusions that users have about the field data, and also to make it easier to use the data outside of our python frontend.

Internally, various quantities like `flux`, `dot` and `poynting` also use colocation to boundaries, which however should not produce any noticeable difference as confirmed by the passing of all the old backend tests. There is a lot more internal refactoring in the associated [backend PR](https://github.com/flexcompute/tidy3d-core/pull/323), which however should not produce any noticeable differences, apart from bugfixes in some edge cases when things like Bloch boundary conditions, symmetries, down-sampling, and / or colocation were not interacting too well with each other.

Probably the most notable refactoring on the frontend is the introduction of a `Simulation.discretize_monitor` method which should always give the exact grid on which field data lives and is also used on the backend. There, the grid is stored as `grid_expanded` when the monitor data is created, since the monitor data doesn't have access to the `simulation` that created it, after it is created. The stored `grid_expanded` is then used not just for expanding symmetries anymore, but also for things like computing dA when integrating flux, as well as computing the coordinates to colocate to if things like `at_boundaries` are called. I was considering making `grid_expanded` a required field, but decided against it for now - but in practice it should always be set whenever the data is created. This design choice was because the previous handling was quite a mess of getting coordinates in three different ways:

a) when expanding symmetries, the coordinates are taken from grid_expanded 
b) when calling sim_data["monitor_name"].at_centers , as also used in e.g. get_poynting and get_intensity, the center coordinates are computed using simulation.discretize(monitor, extend=True) 
c) when calling e.g. MonitorData.poynting or flux for a 2D monitor, coordinates are taken directly from the field data coords. The problem is that there is no guarantee that e.g. all E-field components are present. We could have a MonitorData with just Ex, which is not enough to define the grid boundaries. The reason this had not been a problem yet is I believe largely because we store all field components by default so no one had stumbled on this issue (that we know of)

The present handling unifies all of this into a), and also makes sure that on the backend there is no inconsistency between data coordinates and `grid_expanded` (there used to be some in some edge cases).

tylerflex commented:
A few general comments before I dig into this:
1. I think the unifying of the grid_expanded is a good choice.
2. One thing that worries me regarding backwards compatibility is changing the default behavior of `FieldMonitor.colocate`. If users have scripts (likely) where it is not set explicitly to either True or False, then this PR will change their field outputs. The proper way to handle this probably would be to set `colocate: bool = None` in this version and then warn if `None` that the default has changed from `False` to `True` and to set to `False` if you really want that. However, since field monitors are so widely used, this could generate a lot of warnings, so not sure how to best handle this.  maybe with the log consolidation feature, it won't be so bad? Also, maybe we can display this warning for a single minor release and then drop it later since technically it doesn't need to be corrected. Any thoughts?

momchil-flex commented:
I think that makes sense.

tylerflex reviewed (changes requested):
Looks good! Seems more organized internally and in general I do think this will lead to less confusion. My main concerns are pretty basic:

1. How will this affect the results for people whose scripts used the defaults with fields at yee locations? Are there any edge cases where these changes will break their scripts? Or even where they may get unexpected results? off the top of my head I can't think of any (within the reasonable use case of interpolation inside of the monitor geometry) but just throwing this out there as a consideration.
2. If users interpolate field data to specific points, it will be interpolated twice, no? Is it worth adding a warning, for example, in `FieldData.interp()` if `self.colocate` that it would be more accurate to set `colocate=False` to get the raw data?

tylerflex commented:
Another thing I noticed, not sure if it can be helped really: there is a lot of manual `[1:]` and `[:-1]` indexing happening in the code and I wonder how much of this can be sort of hidden in `Coords` methods rather than being spread around inside of the more higher level logic? Many things are a bit hard to follow because there are often little off-by-one corrections happening inside of the functions and while I don't have any concrete suggestions, wonder if there are some convenience methods that could be employed to make things a bit simpler (like a `Grid.squeeze`, or `Coords.exclude(index=0)` for example)?

momchil-flex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -80,8 +80,9 @@ class AbstractFieldData(MonitorData, AbstractFieldDataset, ABC):
-        description=":class:`.Grid` on which the symmetry will be expanded. "
-        "Required only if any of the ``symmetry`` field are non-zero.",
+        description=":class:`.Grid` discretization of the associated monitor in the simulation "
+        "which created the data. Required if symmetries are present, as "
```
So I was thinking about doing that, but in the end decided against it mostly because for some quick testing it may be convenient not to have to create a grid. Basically there are a few places in our tests, as well as in our docstring examples, where we create a data structure without passing a `grid_expanded`. But maybe this is not a good enough reason and it would be safer to have it required?

tylerflex replied:
up to you. maybe we can warn if not supplied?

momchil-flex replied:
Ok added a warning (also needed to update docstring tests as they erorr if a warning is produced as they expect no output. Not sure if possible to turn off, probably yes, but better have good examples in the docstrings I guess.)

momchil-flex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -174,7 +204,8 @@ class ElectromagneticFieldData(AbstractFieldData, ElectromagneticFieldDataset, A
         if self.monitor.colocate:
-            return self.grid_expanded["centers"]
+            bounds_dict = self.grid_expanded.boundaries.to_dict
+            return Coords(**{key: val[:-1] for key, val in bounds_dict.items()})
```
There are as many fields as there are *cells* in the grid, while the number of grid boundaries is one more than the number of cells. In other words, if the grid is `|_|_|` and monitor.colocate is True, all fields live on the first two boundaries.

Now, thinking about it I realize that it may be nice to just have fields on all three boundaries in that case, but the current way makes things a bit more nicely unified with the `colocate=False` case (where tangential fields live on the first two boundaries, while normal fields live on the first two centers, i.e. they are just recorded in the first two pixels), and is in general kinda assumed in the backend codebase, so I wouldn't change it unless there's a pretty good reason.

momchil-flex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -209,32 +240,52 @@ def _in_plane(self) -> Dict[str, DataArray]:
+    def colocation_boundaries(self) -> Coords:
+        """Coordinates to be used for colocation of the data to grid boundaries."""
+
+        if not self.grid_expanded:
```
This was more like so that a meaningful error will be printed if it was not provided. But yeah the alternative is to make it required and remove this.

momchil-flex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -209,32 +240,52 @@ def _in_plane(self) -> Dict[str, DataArray]:
+        """Coordinates to be used for colocation of the data to grid centers."""
+        colocate_centers = {}
+        for dim, coords in self.colocation_boundaries.to_dict.items():
+            colocate_centers[dim] = (coords[1:] + coords[:-1]) / 2
```
Sorry, what do you mean? `self.colocation_boundaries` is just `Coords` type. You mean adding something like `Coords.centers`?

tylerflex replied:
hm, ok I think I see. yea I guess I was imagining hiding all of these things needing indexing. but I'll leave it up to you. So more generally like `coords.averages` or something.

momchil-flex commented on tidy3d/components/data/sim_data.py:
```diff
@@ -230,23 +230,37 @@ def _at_centers(self, monitor_data: xr.Dataset) -> xr.Dataset:
-        sub_grid = self.simulation.discretize(monitor_data.monitor, extend=False)
-        centers = sub_grid.centers
+        if monitor_data.monitor.colocate:
+            # TODO: this still errors if monitor_data.colocate is allowed to be ``True`` in the
```
Yeah it's probably some sort of casting. However re adding a test, this PR actually adds a validator that `colocate=False` is set for `output_monitors`, because the adjoint source is not correct otherwise. So this comment is more like if in the future `colocate=True` is allowed (which may never happen?). Until then though a unit test doesn't really work as the `colocate=False` validator would have to be turned off. Thoughts?

tylerflex replied:
ok so we dont need to worry about this now I guess. (no need for the test). Regarding colocate=True in adjoint, would be accurate if we added reverse interpolation in the (volumetric) CustomCurrentSource?

momchil-flex replied:
No, you can't actually reverse interpolate back to the original data unambiguously. Some information is lost in the interpolation, the reverse interpolation has to make some assumptions on how to recover the data which are only approximate. The correct thing to do is to backprop through the interpolation. So for example this could work as:

- user makes monitor with colocate = True
- under the hood plugin uses a monitor with colocate = False (so that the adjoint source is correct), and when loading the results makes a colocated `FieldData` (tracked by Jax) to return to user which can then be further manipulated (also tracked by jax)

Then on the backwards pass the interpolation is backpropped through correctly. This can happen because forward data is stored, while reverse interp is agnostic to forward data.

momchil-flex commented:
I've pushed updates based on your comments, thanks! Wherever I didn't comment it just means I implemented your suggestion directly.

> One thing that worries me regarding backwards compatibility is changing the default behavior of FieldMonitor.colocate. If users have scripts (likely) where it is not set explicitly to either True or False, then this PR will change their field outputs. The proper way to handle this probably would be to set colocate: bool = None in this version and then warn if None that the default has changed from False to True and to set to False if you really want that. However, since field monitors are so widely used, this could generate a lot of warnings, so not sure how to best handle this. maybe with the log consolidation feature, it won't be so bad? Also, maybe we can display this warning for a single minor release and then drop it later since technically it doesn't need to be corrected. Any thoughts?

I have added the warning. It may indeed be a bit annoying but I'll rerun the notebooks and see how they look.

> 1. How will this affect the results for people whose scripts used the defaults with fields at yee locations? Are there any edge cases where these changes will break their scripts? Or even where they may get unexpected results? off the top of my head I can't think of any (within the reasonable use case of interpolation inside of the monitor geometry) but just throwing this out there as a consideration.

Yeah not that I can think of. I will rerun all notebooks and will need to see if anything will pop up there...

> 2. If users interpolate field data to specific points, it will be interpolated twice, no? Is it worth adding a warning, for example, in `FieldData.interp()` if `self.colocate` that it would be more accurate to set `colocate=False` to get the raw data?

So `colocate` is actually a `Dataset`-level property so I can't check `monitor.colocate` there directly. I could add a wrapper in `MonitorData`, but it's also used internally in some places and I'm afraid the warning may be produced excessively because of that (although maybe not..) Maybe it's a bit more out of laziness than anything else but it seems to me like a fairly advanced thing to be warning users about.

> Another thing I noticed, not sure if it can be helped really: there is a lot of manual [1:] and [:-1] indexing happening in the code and I wonder how much of this can be sort of hidden in Coords methods rather than being spread around inside of the more higher level logic? Many things are a bit hard to follow because there are often little off-by-one corrections happening inside of the functions and while I don't have any concrete suggestions, wonder if there are some convenience methods that could be employed to make things a bit simpler (like a Grid.squeeze, or Coords.exclude(index=0) for example)?

Yeah honestly I don't really see this helping much, and there are a few ways in which these things are used (e.g. to get sizes like `coords[1:] - coords[:-1]`; to get centers `(coords[1:] + coords[:-1]) / 2`, or just to select a subset because fields are only defined on a subset. Also sometimes this is applied to `Coords` and sometimes to one or more `Coords1D`. I don't see an obvious thing to do that will make things more readable.

momchil-flex commented:
> > 2. If users interpolate field data to specific points, it will be interpolated twice, no? Is it worth adding a warning, for example, in `FieldData.interp()` if `self.colocate` that it would be more accurate to set `colocate=False` to get the raw data?
> 
> So `colocate` is actually a `Dataset`-level property so I can't check `monitor.colocate` there directly. I could add a wrapper in `MonitorData`, but it's also used internally in some places and I'm afraid the warning may be produced excessively because of that (although maybe not..) Maybe it's a bit more out of laziness than anything else but it seems to me like a fairly advanced thing to be warning users about.

I decided to add this warning after all (see latest commit, not sure if you'll approve of the try: except handling). Running notebooks now, will examine if the warning actually appeared anywhere.

tylerflex commented on tidy3d/components/data/dataset.py:
```diff
@@ -73,6 +74,18 @@ def colocate(self, x=None, y=None, z=None) -> xr.Dataset:
         Be sure to apply this method to your field data in those cases.
         """
 
+        try:
```
If we have to do it this way, I'd think that `if hasatttr(self, "monitor"):` is the lesser of two evils.

momchil-flex replied:
The other thing I started doing first was to overwrite the method in MonitorData, but then I have to copy the doc string and it felt kinda ugly too

momchil-flex replied:
Updated.

tylerflex commented on tidy3d/components/monitor.py:
```diff
@@ -32,6 +32,23 @@ class Monitor(Box, ABC):
         min_length=1,
     )
 
+    interval_space: Tuple[Literal[1], ...] = pydantic.Field(
```
I guess it could also be `Literal[(1,1,1)]`?

tylerflex replied:
maybe they need to be floats eg  `1.0`?

momchil-flex replied:
Oh I hadn't forgotten it - it just hasn't updated in this comment view it seems.

But yeah tried again and `Literal[(1,1,1)]` doesn't work (or 1.0 either). I tried various things earlier actually (`Literal[Tuple[1, 1, 1]]` and a few other things) and I could never get it to work, so let's just leave as is now? The error for  `Literal[(1.0,1.0,1.0)]` fyi is 

```
E   pydantic.error_wrappers.ValidationError: 1 validation error for FluxMonitor
E   interval_space
E     unexpected value; permitted: 1.0 (type=value_error.const; given=(1, 1, 1); permitted=(1.0,))
```

tylerflex replied:
yea fine with me

weiliangjin2021 commented on tidy3d/components/monitor.py:
```diff
@@ -32,6 +32,23 @@ class Monitor(Box, ABC):
         min_length=1,
     )
 
+    interval_space: Tuple[Literal[1], Literal[1], Literal[1]] = pydantic.Field(
```
One question: why using "Literal" here rather than pd.PostiveInt?

momchil-flex replied:
That's what it is for monitors in which it can actually be set, it is overwritten [here](https://github.com/flexcompute/tidy3d/blob/3e757803bd9c13bfed3f7eb85df27f88460aaa73/tidy3d/components/monitor.py#L237-L246). `Literal` means that it cannot be anything else than 1. This is how it is for most monitors. It just helps unify things and not have to do things like `if isinstance(...)` when dealing with downsampling.

shashwat-sh commented on tidy3d/plugins/mode/mode_solver.py:
```diff
@@ -96,23 +103,34 @@ def solver_symmetry(self) -> Tuple[Symmetry, Symmetry]:
+        # pylint:disable=protected-access
+        span_inds = self.simulation._discretize_inds_monitor(monitor)
+
+        # Remove extension along monitor normal
```
rather than applying and then removing extensions, would it make sense to parameterize the discretization method to apply or not apply certain extensions?

momchil-flex replied:
Hmmm. Maybe. Currently it is basically driven by the FieldMonitor-s handling. Maybe it's a matter of taste, but I don't like adding kwargs that do some very specific handling that's only used in one other place. But I could think of something like `extend_zero_dims` which can take care of the first two "fixes". And maybe I could also remove the `plane_sym` thing. It should be possible to just discretize the full plane and then take a subset of indexes if symmetries present.

shashwat-sh replied:
Yeah it's a very minor and local thing, so maybe not worth the time

momchil-flex replied:
I ended up completely removing `Simulation.min_sym_box` and reorganizing the related symmetry handling. I had my eye on eliminating that some time ago anyway. But yeah apart from that I left the rest as is.

shashwat-sh commented on tidy3d/components/monitor.py:
```diff
@@ -512,20 +575,17 @@ class ModeSolverMonitor(AbstractModeMonitor):
         "dimension.",
     )
 
+    colocate: bool = pydantic.Field(
```
just curious: to override the `colocate` field in the parent class, why not use a validator to hard code it, instead of making a new field in the subclass?

momchil-flex replied:
I think that way is also possible, I didn't really consider it. One advantage of the current implementation is that it should be directly readable from the docs what the allowed values are.

On the other hand, that does make me wonder: do we really want `colocate` to appear as a Field in all the monitors in which it cannot be modified? My initial approach was to have a property e.g. `colocate_in_solver` that is set for the monitors for which we set it internally, and set by `self.colocate` for monitors where we allow a `colocate` Field. This essentially introduces an extra attribute, but show `colocate` as an input argument to monitors for which it doesn't apply. Same goes for `interval_space` and say `record_in_pml` in the future. @tylerflex suggested the `Literal` approach instead, but does that make the api more confusing?

shashwat-sh replied:
The `Literal` approach does seem most similar to what one might see in a GUI, where the `colocate` field can be "greyed out" and not modifiable for some monitors, but essentially a drop-down menu for others?

shashwat-sh reviewed (commented):
Looks good to me overall, just a couple minor questions.

shashwat-sh commented:
Looks good to me!
